### PR TITLE
Use builder instead of telescoping constructor antipattern in AdminEmailAttributes class #7190

### PR DESCRIPTION
**Pull request by BrianCoveney**
Part of #7190

### CODE REVIEW

**Comment by BrianCoveney** [comment]
Ready for review

**Comment by whipermr5** [comment]
@wkurniawan07 Reminder :)

**Comment by BrianCoveney** [comment]
Will do. I'll refactor the code to match the approach in `StudentProfileAttributes`

**Comment by BrianCoveney** [comment]
Work in progress.

**Comment by BrianCoveney** [comment]
Ready for review

**Comment by wkurniawan07** [comment]
@LimeTheCoder interested to give this one a review?

**Comment by LimeTheCoder** [comment]
@wkurniawan07 sure!

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 1]
```diff
@@ -151,4 +167,43 @@ public String getFirstAddressReceiver() {
...
+        }
+
+        public Builder withEmailId(String emailId) {
+            if (emailId != null) {
+                adminEmailAttributes.emailId = emailId;
+            }
+            return this;
+        }
+
+        public Builder withIsInTrashBin(Boolean isInTrashBin) {
```
Comment (LimeTheCoder) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: Use just boolean here, not wrapper.

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 2]
```diff
@@ -15,37 +15,53 @@
...
-                                List<String> groupReceiver,
-                                Text content,
-                                Date sendDate) {
-        this.subject = subject;
-        this.addressReceiver = addressReceiver;
-        this.groupReceiver = groupReceiver;
-        this.content = content;
-        this.sendDate = sendDate;
+    AdminEmailAttributes() {
+        this.createDate = Const.TIME_REPRESENTS_DEFAULT_TIMESTAMP;
```
Comment (LimeTheCoder) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: 'this' keyword is redundant here

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 3]
```diff
@@ -15,37 +15,53 @@
...
+     * Creates a new AdminEmail with default values for optional fields.
+     *
+     * <p>Following default values are set to corresponding attributes:
+     * <ul>
+     * <li>{@code false} for {@code isInTrashBin}</li>
+     * <li>{@code Const.TIME_REPRESENTS_DEFAULT_TIMESTAMP} for {@code createDate}</li>
+     * <li>{@code Const.ParamsNames.ADMIN_EMAIL_ID} for {@code emailId}</li>
+     * </ul>
+     */
+    public static Builder builder(String subject,
```
Comment (LimeTheCoder) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: move arguments into the same line like in line 174

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 4]
```diff
@@ -151,4 +167,43 @@ public String getFirstAddressReceiver() {
...
         return getGroupReceiver().get(0);
     }
+
+    public static class Builder {
+        private final AdminEmailAttributes adminEmailAttributes;
+
+        public Builder(String subject, List<String> addressReceiver, List<String> groupReceiver,
+                       Text content, Date sendDate) {
+            adminEmailAttributes = new AdminEmailAttributes();
+            adminEmailAttributes.addressReceiver = addressReceiver;
```
Comment (LimeTheCoder) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: Call Assumption.assertNotNull(String message, Object... objects) method here to make sure that all required arguments are present.

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 5]
```diff
@@ -15,37 +15,53 @@
 import teammates.storage.entity.AdminEmail;
 
 public class AdminEmailAttributes extends EntityAttributes<AdminEmail> {
-
-    public String emailId;
+    // Required fields
     public List<String> addressReceiver;
     public List<String> groupReceiver;
     public String subject;
     public Date sendDate;
```
Comment (LimeTheCoder) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: It seems that sendDate field can be optional

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 6]
```diff
@@ -28,8 +28,58 @@
     private String subject = "subject of email";
     private Text content = new Text("valid email content");
     private Date date = new Date();
-    private AdminEmailAttributes adminEmailAttributes = new AdminEmailAttributes(
-            subject, addressReceiverListString, groupReceiverListFileKey, content, date);
+    private AdminEmailAttributes adminEmailAttributes = AdminEmailAttributes
+            .builder(subject, addressReceiverListString, groupReceiverListFileKey, content, date).build();
```
Comment (LimeTheCoder) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: move build() call to the next line

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 7]
```diff
@@ -149,8 +203,13 @@ public void testSanitizeForSaving() {
...
         ______TS("valid sanitation of admin email");
 
-        AdminEmailAttributes adminEmailAttributes = new AdminEmailAttributes(
-                subjectWithWhitespaces, addressReceiverListString, groupReceiverListFileKey, contentWithWhitespaces, date);
+        AdminEmailAttributes adminEmailAttributes = AdminEmailAttributes
+                .builder(subjectWithWhitespaces,
+                        addressReceiverListString,
+                        groupReceiverListFileKey,
+                        contentWithWhitespaces,
+                        date)
```
Comment (LimeTheCoder) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: no need for the new line for each argument

File: src/main/java/teammates/storage/entity/AdminEmail.java [hunk 8]
```diff
@@ -63,6 +62,19 @@ public AdminEmail(List<String> addressReceiver, List<String> groupReceiver, Stri
         this.isInTrashBin = false;
     }
 
+    // This constructor is used only for AdminEmailAttributesTest#testValueOf()
+    public AdminEmail(Long emailId, List<String> addressReceiver, List<String> groupReceiver, String subject,
```
Comment (LimeTheCoder) [inline src/main/java/teammates/storage/entity/AdminEmail.java]: It's not good to add new public constructor only for one test purpose. Instead, try to avoid null pointer exception in getEmailId() method:
```java
    public String getEmailId() {
        return emailId == null ? null : Key.create(AdminEmail.class, this.emailId).toWebSafeString();
    }
```

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java (hunk 1 above)
Comment (BrianCoveney) [reply src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: We cannot check for null with the primitive boolean type. 
Therefore, .withIsInTrashBin(null) is a compilation error.
I could ommit the test for null here, but I don't think this is a good idea. What do you think?

**Comment by BrianCoveney** [comment]
Ready for review.

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 9]
```diff
@@ -15,37 +16,52 @@
...
-                                List<String> groupReceiver,
-                                Text content,
-                                Date sendDate) {
-        this.subject = subject;
-        this.addressReceiver = addressReceiver;
-        this.groupReceiver = groupReceiver;
-        this.content = content;
-        this.sendDate = sendDate;
+    AdminEmailAttributes() {
+        sendDate = null;
```
Comment (LimeTheCoder) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: Redundant assignment, sendDate by default is null

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 10]
```diff
@@ -15,37 +16,52 @@
...
+        isInTrashBin = false;
+    }
+
+    /**
+     * Creates a new AdminEmail with default values for optional fields.
+     *
+     * <p>Following default values are set to corresponding attributes:
+     * <ul>
+     * <li>{@code false} for {@code isInTrashBin}</li>
+     * <li>{@code null} for {@code sendDate}</li>
```
Comment (LimeTheCoder) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: Remove it too, since null is default value for all uninitiliazed variables

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 11]
```diff
@@ -151,4 +167,52 @@ public String getFirstAddressReceiver() {
...
+        private final AdminEmailAttributes adminEmailAttributes;
+
+        public Builder(String subject, List<String> addressReceiver, List<String> groupReceiver, Text content) {
+            adminEmailAttributes = new AdminEmailAttributes();
+            adminEmailAttributes.addressReceiver = addressReceiver;
+            adminEmailAttributes.groupReceiver = groupReceiver;
+            adminEmailAttributes.subject = subject;
+            adminEmailAttributes.content = content;
+
+            Assumption.assertNotNull("Non-null value expected for required AdminEmailAttributes",
```
Comment (LimeTheCoder) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: move assumption check before adminEmailAttributes creation, e.g before line 175

File: src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java [hunk 12]
```diff
@@ -61,12 +61,11 @@ protected ActionResult execute() {
...
         if (isError) {
-            data.emailToEdit = new AdminEmailAttributes(subject,
-                                                        addressReceiver,
-                                                        groupReceiver,
-                                                        new Text(emailContent),
-                                                        null);
-            data.emailToEdit.emailId = emailId;
+            data.emailToEdit = AdminEmailAttributes
+                    .builder(subject, addressReceiver, groupReceiver, new Text(emailContent))
+                    .withSendDate(null)
```
Comment (LimeTheCoder) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java]: redundant

File: src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java [hunk 13]
```diff
@@ -83,11 +82,10 @@ private void updatePreviousEmailDraft(String previousEmailId,
...
                                           ) {
 
-        AdminEmailAttributes newDraft = new AdminEmailAttributes(subject,
-                                                                 addressReceiver,
-                                                                 groupReceiver,
-                                                                 new Text(content),
-                                                                 null);
+        AdminEmailAttributes newDraft = AdminEmailAttributes
+                .builder(subject, addressReceiver, groupReceiver, new Text(content))
+                .withSendDate(null)
```
Comment (LimeTheCoder) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java]: remove it too

File: src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java [hunk 14]
```diff
@@ -102,11 +100,10 @@ private void createAndSaveNewDraft(String subject,
...
                                        String content) {
 
-        AdminEmailAttributes newDraft = new AdminEmailAttributes(subject,
-                                                                 addressReceiver,
-                                                                 groupReceiver,
-                                                                 new Text(content),
-                                                                 null);
+        AdminEmailAttributes newDraft = AdminEmailAttributes
+                .builder(subject, addressReceiver, groupReceiver, new Text(content))
+                .withSendDate(null)
```
Comment (LimeTheCoder) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSaveAction.java]: remove it too

File: src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java [hunk 15]
```diff
@@ -74,12 +74,12 @@ protected ActionResult execute() {
...
         if (isError) {
-            data.emailToEdit = new AdminEmailAttributes(subject,
-                                                        addressReceiver,
-                                                        groupReceiver,
-                                                        new Text(emailContent),
-                                                        null);
-            data.emailToEdit.emailId = emailId;
+            data.emailToEdit = AdminEmailAttributes
+                    .builder(subject, addressReceiver, groupReceiver, new Text(emailContent))
+                    .withSendDate(null)
```
Comment (LimeTheCoder) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java]: remove it too

File: src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java [hunk 16]
```diff
@@ -92,12 +92,11 @@ protected ActionResult execute() {
...
```
Comment (LimeTheCoder) [inline src/main/java/teammates/ui/controller/AdminEmailComposeSendAction.java]: no need to set null here, since sendDate by default null

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 17]
```diff
@@ -28,8 +28,77 @@
...
+        try {
+            AdminEmailAttributes attributesWithNullRequiredArguments = AdminEmailAttributes
+                        .builder(null, null, null, null)
+                        .build();
+
+            attributesWithNullRequiredArguments.isValid();
+            signalFailureToDetectException(" - AssertionError");
+        } catch (AssertionError ae) {
+            assertEquals("Non-null value expected for required AdminEmailAttributes", ae.getMessage());
+        }
```
Comment (LimeTheCoder) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: No need here for try-catch block. You can add something like that
`@Test(expectedExceptions = AssertionError.class, expectedExceptionsMessageRegExp = "Non-null value expected for required AdminEmailAttributes")` annotation on this method

**Comment by BrianCoveney** [comment]
Ready for review.

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 18]
```diff
@@ -22,14 +22,82 @@
...
+        ______TS("success: default values for optional params");
+
+        assertEquals(Const.ParamsNames.ADMIN_EMAIL_ID, attributesWithNullOptionalArguments.getEmailId());
+        assertFalse("Default false for isInTrashBin", attributesWithNullOptionalArguments.isInTrashBin);
+        assertEquals(Const.TIME_REPRESENTS_DEFAULT_TIMESTAMP, attributesWithNullOptionalArguments.getCreateDate());
+        assertEquals(null, attributesWithNullOptionalArguments.getSendDate());
+    }
+
+    @Test
+    public void testBuilderWithNullRequiredArguments() {
```
Comment (LimeTheCoder) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: Use here expectedExceptions and expectedExceptionsMessageRegExp annotation args

File: src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java [hunk 19]
```diff
@@ -253,27 +254,34 @@ public void testExecuteAndPostProcess() throws Exception {
...
+        try {
+            emailId = "nonExisitingId";
+            content = "valid content";
+            subject = "";
+            receiver = "test@example.tmt";
+            action = getAction(
+                    Const.ParamsNames.ADMIN_EMAIL_CONTENT, content,
+                    Const.ParamsNames.ADMIN_EMAIL_SUBJECT, subject,
+                    Const.ParamsNames.ADMIN_EMAIL_ADDRESS_RECEIVERS, receiver,
+                    Const.ParamsNames.ADMIN_EMAIL_ID, emailId);
```
Comment (LimeTheCoder) [inline src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java]: It seems that exception will occur in that line and all lines above will not executed. I think its better to not add AssertNotEmpty to builder constructor since we have other mechanism to handle such cases.

**Comment by BrianCoveney** [comment]
Ready for review.

**Comment by BrianCoveney** [comment]
Hi @wkurniawan07 when you get the chance, can we get a final review on this PR?

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 20]
```diff
@@ -15,37 +16,49 @@
...
-                                Date sendDate) {
-        this.subject = subject;
-        this.addressReceiver = addressReceiver;
-        this.groupReceiver = groupReceiver;
-        this.content = content;
-        this.sendDate = sendDate;
+    AdminEmailAttributes() {
+        createDate = Const.TIME_REPRESENTS_DEFAULT_TIMESTAMP;
+        emailId = Const.ParamsNames.ADMIN_EMAIL_ID;
+        isInTrashBin = false;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: This is unnecessary as default `boolean` value in Java is `false`.

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 21]
```diff
@@ -15,37 +16,49 @@
...
+    }
+
+    /**
+     * Creates a new AdminEmail with default values for optional fields.
+     *
+     * <p>Following default values are set to corresponding attributes:
+     * <ul>
+     * <li>{@code false} for {@code isInTrashBin}</li>
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: But keep this.

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 22]
```diff
@@ -15,37 +16,49 @@
...
+    public static Builder builder(String subject, List<String> addressReceiver, List<String> groupReceiver, Text content) {
+        return new Builder(subject, addressReceiver, groupReceiver, content);
+    }
+
+    public static AdminEmailAttributes valueOf(AdminEmail adminEmail) {
+        return new Builder(
+                adminEmail.getSubject(),
+                adminEmail.getAddressReceiver(),
+                adminEmail.getGroupReceiver(),
+                adminEmail.getContent())
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: Looks not really easy to read. This might be better:
```java
return new Builder(adminEmail.getSubject(), adminEmail.getAddressReceiver(),
                adminEmail.getGroupReceiver(), adminEmail.getContent())
        .with*(*)
        .build();
```

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java (hunk 1 above)
Comment (wkurniawan07) [reply src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: Why do you need to test for `null` if passing `null` to this method is a compile-time error in the first place?

File: src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java [hunk 23]
```diff
@@ -265,7 +265,6 @@ public void testExecuteAndPostProcess() throws Exception {
         pageResult = getShowPageResult(action);
         assertEquals(getPageResultDestination(Const.ViewURIs.ADMIN_EMAIL, true, "admin.user"),
                 pageResult.getDestinationWithParams());
-
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/action/AdminEmailComposeSaveActionTest.java]: Revert this?

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 24]
```diff
@@ -22,14 +22,75 @@
...
+        assertEquals(Const.ParamsNames.ADMIN_EMAIL_ID, attributesWithNullOptionalArguments.getEmailId());
+        assertFalse("Default false for isInTrashBin", attributesWithNullOptionalArguments.isInTrashBin);
+        assertEquals(Const.TIME_REPRESENTS_DEFAULT_TIMESTAMP, attributesWithNullOptionalArguments.getCreateDate());
+        assertEquals(null, attributesWithNullOptionalArguments.getSendDate());
+    }
+
+    @Test(expectedExceptions = AssertionError.class, expectedExceptionsMessageRegExp = Const.StatusCodes.NULL_PARAMETER)
+    public void testBuilderWithNullRequiredArguments() {
+        AdminEmailAttributes
+                .builder(null, null, null, null)
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: You should test all the `null`s individually.

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 25]
```diff
@@ -103,28 +168,18 @@ public void testValidate() throws Exception {
         assertFalse("Invalid input", attributes.isValid());
     }
 
-    @Test
+    @Test(expectedExceptions = AssertionError.class, expectedExceptionsMessageRegExp = Const.StatusCodes.NULL_PARAMETER)
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: This should be covered under `testBuilderWithNullRequiredArguments()`?

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 26]
```diff
@@ -103,28 +168,18 @@ public void testValidate() throws Exception {
...
-        } catch (AssertionError e) {
-            ignoreExpectedException();
-        }
+        AdminEmailAttributes
+                .builder(subject, addressReceiverListString, groupReceiverListFileKey, null)
+                .build();
     }
 
-    @Test
+    @Test(expectedExceptions = AssertionError.class, expectedExceptionsMessageRegExp = Const.StatusCodes.NULL_PARAMETER)
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: This should be covered under `testBuilderWithNullRequiredArguments()`?

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 27]
```diff
@@ -22,14 +22,75 @@
...
 
-    private FieldValidator fieldValidator = new FieldValidator();
     private List<String> addressReceiverListString = Arrays.asList("example1@test.com", "example2@test.com");
     private List<String> groupReceiverListFileKey = Arrays.asList("listfilekey", "listfilekey");
     private String subject = "subject of email";
     private Text content = new Text("valid email content");
     private Date date = new Date();
-    private AdminEmailAttributes adminEmailAttributes = new AdminEmailAttributes(
-            subject, addressReceiverListString, groupReceiverListFileKey, content, date);
+    private AdminEmailAttributes adminEmailAttributes = AdminEmailAttributes
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: I can't seem to find it; what do you need this variable for?

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java (hunk 21 above)
Comment (BrianCoveney) [reply src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: As we are keeping Javadoc for ```isInTrashBin```, should we also keep it for ```sendDate```?  
```<li>{@code null} for {@code sendDate}</li>```

Comment (wkurniawan07) [reply src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: @BrianCoveney sounds like a good idea.

File: src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java [hunk 28]
```diff
@@ -22,37 +22,121 @@
...
+
+    @Test(expectedExceptions = AssertionError.class, expectedExceptionsMessageRegExp = Const.StatusCodes.NULL_PARAMETER)
+    public void testBuilderWithNullRequiredArguments() {
+        ______TS("failure: subject cannot be null)");
+
+        AdminEmailAttributes
+                .builder(null, addressReceiverListString, groupReceiverListFileKey, content)
+                .build();
+
+        ______TS("failure: addressReceiverListString cannot be null)");
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/datatransfer/AdminEmailAttributesTest.java]: I don't think this is a good idea; the first instance of `*Exception` thrown will already make this test pass, which means everything past this line will not be executed.

**Comment by BrianCoveney** [comment]
Ready for review

File: src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java [hunk 29]
```diff
@@ -15,37 +16,46 @@
...
     public List<String> addressReceiver;
     public List<String> groupReceiver;
     public String subject;
+    public Text content;
+
+    // Optional fields
     public Date sendDate;
     public Date createDate;
-    public Text content;
+    public String emailId;
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/attributes/AdminEmailAttributes.java]: Why did this become optional?

File: src/main/java/teammates/storage/entity/AdminEmail.java [hunk 30]
```diff
@@ -88,7 +87,7 @@ public void setSendDate(Date sendDate) {
     }
 
     public String getEmailId() {
-        return Key.create(AdminEmail.class, this.emailId).toWebSafeString();
+        return emailId == null ? null : Key.create(AdminEmail.class, this.emailId).toWebSafeString();
```
Comment (whipermr5) [inline src/main/java/teammates/storage/entity/AdminEmail.java]: When is this null?

Comment (wkurniawan07) [reply src/main/java/teammates/storage/entity/AdminEmail.java]: Many instances in production code (not just test code, otherwise I'll also complain) don't require this attribute to be equipped with `emailId`. So it is already optional to begin with.

Use case: if you want to pass it to functions that don't require the existence of `emailId`.

It may not be the best deciding factor on what to count a field as "optional" or not, but if you really want to pick on that, I'll suggest you relook at all the attributes builders instead of just picking on this one.
